python_src/program_markers/instrumenter.py

```diff
@@ -196,27 +198,87 @@ def get_constant_macro(self) -> str:
 
-    The markers are found by searching for calls or jumps to functions with
-    names starting with a known marker prefix (e.g., DCEMarker123_)
+    Subclasses of MarkerStrategy are used to specify what kind of marker preprocessor
+    directives should be generated by `Instrumenter.with_marker_directives`.
+    """
+
+    @staticmethod
+    @abstractmethod
+    def name() -> str:
+        """Returns the name of the marker strategy
+
+        Returns:
+            str:
+                the name
+        """
+        raise NotImplementedError
+
+    @staticmethod
+    @abstractmethod
+    def make_preprocessor_code(marker: DCEMarker) -> str:
+        """Generates the preprocessor code of this strategy for the marker
+
+        Args:
+            marker (DCEMarker):
+                the marker for which the preprocessor code should be generated
+
+        Returns:
+            str:
+                the marker preprocessor code
+        """
+        raise NotImplementedError
+
+    @staticmethod
+    @abstractmethod
+    def marker_detection_regex() -> re.Pattern[str]:
+        """Returns a pattern of the regex used to detect markers of this strategy
+
+        Returns:
+            re.Pattern[str]:
+                the regex pattern
+        """
+        raise NotImplementedError
+
+    @staticmethod
+    @abstractmethod
+    def regex_marker_id_group_index() -> int:
+        """Returns the index of the group in the regex, generated by
+        marker_detection_regex(), containing the marker id
+
+        Returns:
+            int:
+                the index of the id group
+        """
+        raise NotImplementedError
+
+
+def find_non_eliminated_markers_impl(
+    asm: str, marker_strategy: MarkerStrategy
+) -> tuple[Marker, ...]:
+    """Finds non-eliminated markers of the type specified by `marker_strategy` in
+    `asm`.
+
+    The DCE markers are found by matching the regex pattern specified by the given
+    strategy.
+    The VR markers are found by searching for calls or jumps to functions with
+    names starting with a known marker prefix (e.g., VRMarkerE123_)
 
     Returns:
         tuple[Marker, ...]:
             The non-eliminated markers for the given compilation setting.
     """
-    dce_non_eliminated_regex = re.compile(
-        f".*[call|jmp].*{DCEMarker.prefix}([0-9]+)_.*"
-    )
     vr_non_eliminated_regex = re.compile(
         f".*[call|jmp].*{VRMarker.prefix}([G|L])E([0-9]+)_.*"
     )
     non_eliminated_markers: set[Marker] = set()
     for line in asm.split("\n"):
-        if m := dce_non_eliminated_regex.match(line.strip()):
-            non_eliminated_markers.add(DCEMarker(f"{DCEMarker.prefix}{m.group(1)}_"))
-        elif m := vr_non_eliminated_regex.match(line.strip()):
+        if m := marker_strategy.marker_detection_regex().match(line.strip()):
+            idx = marker_strategy.regex_marker_id_group_index()
+            non_eliminated_markers.add(DCEMarker(f"{DCEMarker.prefix}{m.group(idx)}_"))
```

Comment:
Last comment, I am wondering if it make sense to provide a method in `MarkerStrategy` that does what the above lines do:
`def detect_marker(asm_line: str) -> DCEMarker | None`

Reply:
Yes I also think it makes sense to have such a convenience method and I just added it.